### PR TITLE
[libc] Lock the output stream for the 'puts' call

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
The `puts` function consists of an initial write and then another write
to append the newline. When executing code in parallel, it is possible
for these writes to becomes disjointed. This code adds an explicit lock
call to ensure that the string is always appended by the newline as the
users expects.

Wasn't sure if this required a test as it would be difficult since
reproducing it would be flaky.

### CODE REVIEW

File: libc/src/stdio/generic/puts.cpp [hunk 1]
```diff
@@ -15,8 +15,21 @@
...
+// Simple helper to unlock the file once destroyed.
+struct ScopedLock {
+  ScopedLock(LIBC_NAMESPACE::File *stream) : stream(stream) { stream->lock(); }
+  ~ScopedLock() { stream->unlock(); }
+
+private:
+  LIBC_NAMESPACE::File *stream;
+};
+
 LLVM_LIBC_FUNCTION(int, puts, (const char *__restrict str)) {
```
Comment (SchrodingerZhu) [inline libc/src/stdio/generic/puts.cpp:27]: why this is marked as `__restrict`?

Comment (jhuber6) [reply libc/src/stdio/generic/puts.cpp:27]: I think all of the implementations of string functions are like this. My best guess is that it's giving the same effect as `const char const *` without changing the signature or something.

**Comment by jhuber6** [comment]
ping

**Review by lntue — APPROVED** [review]

File: libc/src/stdio/generic/puts.cpp (hunk 1 above)
Comment (nickdesaulniers) [reply libc/src/stdio/generic/puts.cpp:31]: Sorry, can you please provide more info on this point; it doesn't sound right to me.

My understanding of `__restrict` is that it has to do with strict aliasing of function parameters. If there's only 1 parameter, there is __nothing__ to be concerned with wrt to aliasing.

So why add __restrict here?

**Pull request merged**